src/metabase/api/x_ray.clj

```diff
@@ -32,163 +33,133 @@
+  [query]
+  (->> {:dataset_query query}
+       (merge (card/query->database-and-table-ids query))
+       card/map->CardInstance))
 
-(api/defendpoint GET "/card/:id"
-  "Get x-ray for a `Card` with ID."
-  [id max_query_cost max_computation_cost]
+(api/defendpoint POST "/query"
+  "Get x-ray of a query with no associated model."
+  [max_query_cost max_computation_cost :as {{:keys [database], :as query} :body}]
   {max_query_cost       MaxQueryCost
    max_computation_cost MaxComputationCost}
   (api/check-403 (costs/enable-xrays))
-  (->> id
-       (api/read-check Card)
+  (when-not (= database database/virtual-id)
+    (api/read-check Database database))
+  (->> query
+       virtual-card
        (fe/extract-features {:max-cost (max-cost max_query_cost
                                                  max_computation_cost)})
        fe/x-ray))
 
-(api/defendpoint GET "/compare/fields/:id1/:id2"
-  "Get comparison x-ray for `Field`s with ID1 and ID2."
-  [id1 id2 max_query_cost max_computation_cost]
+(api/defendpoint GET "/compare/:models/:id1/:id2"
+  "Get comparison x-ray for two models of type `:models` with ids :id1 and id2."
+  [models id1 id2 max_query_cost max_computation_cost]
   {max_query_cost       MaxQueryCost
-   max_computation_cost MaxComputationCost}
+   max_computation_cost MaxComputationCost
+   models               Model}
   (api/check-403 (costs/enable-xrays))
   (->> [id1 id2]
-       (map #(api/read-check Field (Integer/parseInt %)))
+       (map #(api/read-check (->model models) (Long/parseLong %)))
        (apply fe/compare-features
               {:max-cost (max-cost max_query_cost max_computation_cost)})
        fe/x-ray))
 
-(api/defendpoint GET "/compare/tables/:id1/:id2"
-  "Get comparison x-ray for `Table`s with ID1 and ID2."
-  [id1 id2 max_query_cost max_computation_cost]
+(api/defendpoint GET "/compare/:model1/:id1/model2/:id2"
+  "Get comparison x-ray for two models of types `:model1` and `:model2` with
+   ids `:id1` and `:id2`."
+  [model1 model2 id1 id2 max_query_cost max_computation_cost]
   {max_query_cost       MaxQueryCost
-   max_computation_cost MaxComputationCost}
+   max_computation_cost MaxComputationCost
+   model1               Model
+   model2               Model}
   (api/check-403 (costs/enable-xrays))
-  (->> [id1 id2]
-       (map #(api/read-check Table (Integer/parseInt %)))
-       (apply fe/compare-features
-              {:max-cost (max-cost max_query_cost max_computation_cost)})
-       fe/x-ray))
+  (fe/x-ray
+   (fe/compare-features
+    {:max-cost (max-cost max_query_cost max_computation_cost)}
+    (api/read-check (->model model1) (Long/parseLong id1))
+    (api/read-check (->model model2) (Long/parseLong id2)))))
 
-(api/defendpoint GET "/compare/tables/:id1/:id2/field/:field"
-  "Get comparison x-ray for `Field` named `field` from `Table`s with ID1 and
-   ID2."
-  [id1 id2 field max_query_cost max_computation_cost]
+(api/defendpoint GET "/compare/:model/:id1/:id2/field/:field"
```

Comment:
I feel like this is kind of backwards. Why are we passing in the underlying table/segment IDs here rather than just the two field ids? Again, I think the whole "everything is a generic data driven method" thing is obscuring a simpler specific view of the comparison.

Reply:
This is actually a very good example of what you were cautioning about. For tables it's not needed, for segments however you do need to specify them, as you want to xray only the field as it's seen through the segment.